### PR TITLE
feat: add JoinedDate to User

### PR DESCRIPTION
**Pull request by WiktorSinkiewicz**
Closes #86

## Changes
- Added `JoinedDate` property (`DateTimeOffset` type) to the `User` domain entity.
- Ensured the property is immutable (no setter / using `init`).
- Exposed the `JoinedDate` field in the Data Transfer Objects (`UserDto`).
- Generated a new EF Core database migration (`AddUserJoinedDate`).

## How to test (optional)
1. Run the database (`docker compose up database`) and the backend server (`dotnet run`).
2. Navigate to the interactive API documentation (e.g., `http://localhost:7278/docs`).
3. Scroll down to the "Models" section and check the schema for `UserDto` - the `joinedDate` field should be visible.

## Screenshots / recordings (for UI stuff)
...

## Checklist
- [x] PR is linked to an issue (tab on the right).
- [x] Acceptance criteria (from issue) are met.
- [x] All status checks (CI) are green.
- [ ] Tests added / updated.
- [ ] Docs updated (if applicable).

---
To have your PR reviewed put the link e.g. `https://github.com/akai-org/put-wiki/pull/0` to the `Review PR` thread on [put-wiki dc channel](https://discord.com/channels/768494845634412624/1467612224079007855) (you must be member of the AKAI discord server)

### CODE REVIEW

File: server/Application/DTOs/UserDto.cs [hunk 1]
```diff
@@ -1,3 +1,4 @@
-﻿namespace Application.DTOs;
+﻿using System;
+namespace Application.DTOs;
 
-public record UserDto(string Id, string HashedUsosId);
\ No newline at end of file
+public record UserDto(string Id, string HashedUsosId, DateTimeOffset JoinedDate);
```
Comment (Aleks334) [inline server/Application/DTOs/UserDto.cs:4]: Here I thought about using just string, but after testing it manually, it correctly converts it using built-in JSON serializer for asp.net core. So that's fine

File: server/Domain/Users/User.cs [hunk 2]
```diff
@@ -14,5 +15,6 @@ public User(string hashedUsosId)
 
         Id = Guid.CreateVersion7();
         HashedUsosId = hashedUsosId;
+        JoinedDate = DateTimeOffset.UtcNow;
```
Comment (Aleks334) [inline server/Domain/Users/User.cs]: Here's a small catch. If you ever wanted to test this class for some case related to time, e.g. user can post opinion after X hours from account creation. Then to unit test this you need to manipulate time somehow. Having `DateTimeOffset.UtcNow` hardcoded inside the User entity constructor will make it impossible to mock. You need to inject joined date to constructor of this class. It is instantiated in `ProvisionUserUseCase` in application layer. 

You can enhance it even more by injecting built-in `TimeProvider` in use case class to get the current time and pass it down. This acts as a system clock that we can easily fake in our future unit tests using `FakeTimeProvider`.

See this article: https://andrewlock.net/exploring-the-dotnet-8-preview-avoiding-flaky-tests-with-timeprovider-and-itimer/

**Review by Aleks334 — CHANGES_REQUESTED** [review]
Good job! There's only one thing that can be improved

File: server/PutWiki.UnitTests/Application/ProvisionUserUseCaseTests.cs [hunk 3]
```diff
@@ -134,7 +134,7 @@ public async Task ExecuteAsync_WhenUserDoesNotExistInDb_ShouldCreateNewUserAndSa
         _userRepositoryMock
             .Setup(x => x.GetByHashedUsosIdAsync(hashedUsosId, It.IsAny<CancellationToken>()))
             .ReturnsAsync((User?)null);
-        
+
```
Comment (Aleks334) [inline server/PutWiki.UnitTests/Application/ProvisionUserUseCaseTests.cs:138]: btw, make sure that you have precommit git hook configured so that `dotnet format` runs automatically

File: server/PutWiki.UnitTests/Application/ProvisionUserUseCaseTests.cs [hunk 4]
```diff
@@ -27,13 +28,15 @@ public class ProvisionUserUseCaseTests
     private readonly Mock<IUsosOAuthService> _usosOAuthServiceMock;
     private readonly Mock<IUsosIdHasher> _idHasherMock;
     private readonly Mock<IUserRepository> _userRepositoryMock;
+    private readonly FakeTimeProvider _timeProviderMock;
```
Comment (Aleks334) [inline server/PutWiki.UnitTests/Application/ProvisionUserUseCaseTests.cs]: here, you added `Mock` suffix but in reality this is fake, not mock object. There are several types of test doubles, not only mocks.
See for example this article: https://martinfowler.com/articles/mocksArentStubs.html#TheDifferenceBetweenMocksAndStubs

File: server/PutWiki.UnitTests/Application/ProvisionUserUseCaseTests.cs [hunk 5]
```diff
@@ -138,6 +144,7 @@ public async Task ExecuteAsync_WhenUserDoesNotExistInDb_ShouldCreateNewUserAndSa
         result.Value.Should().NotBeNull();
         result.Value!.HashedUsosId.Should().Be(hashedUsosId);
         result.Value.Id.Should().NotBe(Guid.Empty.ToString());
+        result.Value.JoinedDate.Should().Be(_timeProviderMock.GetUtcNow());
```
Comment (Aleks334) [inline server/PutWiki.UnitTests/Application/ProvisionUserUseCaseTests.cs]: here you assert specific value from fake object. Usually for predictability you should compare the result with predicted constant value. Like above `hashedUsosId` from `Arrange` step in this test.

File: server/PutWiki.UnitTests/Application/ProvisionUserUseCaseTests.cs [hunk 6]
```diff
@@ -138,6 +144,7 @@ public async Task ExecuteAsync_WhenUserDoesNotExistInDb_ShouldCreateNewUserAndSa
...
 
         _userRepositoryMock.Verify(x => x.Add(It.Is<User>(u => u.HashedUsosId == hashedUsosId)), Times.Once);
```
Comment (Aleks334) [inline server/PutWiki.UnitTests/Application/ProvisionUserUseCaseTests.cs]: Besides asserting only DTO above, you can also verify, whether domain entity (`User`) is passed to repository with correct date (another comparison is needed like for hashedUsosId). Then you will test both use case output (DTO) and side effect (saving to repo)

File: server/Presentation/Program.cs [hunk 7]
```diff
@@ -21,6 +23,8 @@
 builder.Services.AddHealthChecks();
 builder.Services.AddMemoryCache();
 
+builder.Services.AddSingleton(TimeProvider.System);
```
Comment (Aleks334) [inline server/Presentation/Program.cs]: For better readability we decided not to register everything directly inside `Program.cs`. Therefore please put this line to dedicated extension method for one of layers. 

I think that like file system is more infrastructure detail, similarly system clock is also suitable for it (`AddInfrastructure()` below)

**Review by Aleks334 — CHANGES_REQUESTED** [review]
I found several small issues in updated tests. Also please make sure that each phase (arrange, act, assert) in tests is separated by whitespace from others. This improves readability a lot.

File: server/PutWiki.UnitTests/Application/ProvisionUserUseCaseTests.cs (hunk 3 above)
Comment (WiktorSinkiewicz) [reply server/PutWiki.UnitTests/Application/ProvisionUserUseCaseTests.cs:138]: Thanks, it should work now

File: server/PutWiki.UnitTests/Application/ProvisionUserUseCaseTests.cs (hunk 4 above)
Comment (WiktorSinkiewicz) [reply server/PutWiki.UnitTests/Application/ProvisionUserUseCaseTests.cs]: Thanks, now I think I know what those mean

File: server/PutWiki.UnitTests/Application/ProvisionUserUseCaseTests.cs (hunk 6 above)
Comment (WiktorSinkiewicz) [reply server/PutWiki.UnitTests/Application/ProvisionUserUseCaseTests.cs]: Done, should check properly before saving

File: server/Presentation/Program.cs (hunk 7 above)
Comment (Aleks334) [reply server/Presentation/Program.cs]: @WiktorSinkiewicz It seems that you forgot to remove TimeProvider registration from Program.cs after adding to AddInfrastructure()

Comment (WiktorSinkiewicz) [reply server/Presentation/Program.cs]: 😭🥀

**Review by Aleks334 — APPROVED** [review]
Your first PR approved, congrats!
